File: python/taflow/metrics/breakeven_rate.py
"""Exact-zero observation rate metric."""
from __future__ import annotations
from typing import Any
from .._native.metrics import BreakevenRate as _Native
from ._input import as_metric_series


class BreakevenRate:
    """Compute the fraction of valid observations exactly equal to zero.

    The independent oracle is NumPy exact-zero counting. Strictly positive
    observations are wins, strictly negative observations are losses, and both
    positive and negative floating zero are breakeven. Warm-up requires one
    usable observation. Inputs are explicitly period returns, raw period P&L,
    or closed-trade P&L; no annualization or capital conversion is performed.
    Rust owns the allocation-free O(1) state and NaNs are omitted by default.
    """
    def __init__(self, nan_policy: str = "omit") -> None:
        """Initialize an empty configured metric."""
        self._state = _Native(nan_policy)

    def from_returns(
        self, returns: Any, *, column: str | None = None
    ) -> "BreakevenRate":
        """Append chronological returns observations and return this metric."""
        self._state.from_returns(as_metric_series(returns, column=column))
        return self

    def from_pnl(
        self, pnl: Any, *, column: str | None = None
    ) -> "BreakevenRate":
        """Append chronological pnl observations and return this metric."""
        self._state.from_pnl(as_metric_series(pnl, column=column))
        return self

    def from_trades(
        self, trades: Any, *, column: str | None = None
    ) -> "BreakevenRate":
        """Append chronological trades observations and return this metric."""
        self._state.from_trades(as_metric_series(trades, column=column))
        return self

    def append(self, value: float) -> "BreakevenRate":
        """Append one observation and return this metric."""
        self._state.append(float(value)); return self
    def extend(self, values: Any, *, column: str | None = None) -> "BreakevenRate":
        """Append observations and return this metric."""
        self._state.extend(as_metric_series(values, column=column)); return self
    @property
    def value(self) -> float | None:
        """Return the breakeven fraction, or ``None`` during warm-up."""
        return self._state.value
    def compute(self) -> float | None:
        """Return the current scalar without replaying input."""
        return self._state.compute()
    def reset(self) -> "BreakevenRate":
        """Clear observations, preserve the domain, and return this metric."""
        self._state.reset(); return self
    def __len__(self) -> int:
        """Return valid observation count delegated to Rust."""
        return len(self._state)


__all__ = ["BreakevenRate"]

File: crates/taflow-metrics/src/metrics/breakeven_rate.rs
use crate::{MetricError, MetricInputKind, MetricInputState, MetricResult, NanPolicy};

/// Fraction of valid observations exactly equal to zero.
#[derive(Debug, Clone)]
pub struct BreakevenRate {
    input: MetricInputState,
    breakevens: usize,
}

impl BreakevenRate {
    /// Construct an empty state for returns, raw period P&L, or closed trades.
    pub fn new(nan_policy: NanPolicy) -> MetricResult<Self> {
        Ok(Self {
            input: MetricInputState::unbound(nan_policy),
            breakevens: 0,
        })
    }
    /// Append one observation and return the fraction exactly equal to zero.
    pub fn append(&mut self, value: f64) -> MetricResult<Option<f64>> {
        if let Some(observation) = self.input.append(value)? {
            self.breakevens += usize::from(observation == 0.0);
        }
        Ok(self.value())
    }
    /// Append a slice through the same persistent state.
    pub fn extend(&mut self, values: &[f64]) -> MetricResult<Option<f64>> {
        self.input.extend(values, |observation| {
            self.breakevens += usize::from(observation == 0.0);
            Ok(())
        })?;
        Ok(self.value())
    }
    /// Return exact-zero count divided by valid observation count.
    pub fn value(&self) -> Option<f64> {
        (!self.input.is_empty()).then(|| self.breakevens as f64 / self.input.len() as f64)
    }
    /// Return the current scalar without replaying input.
    pub fn compute(&self) -> Option<f64> {
        self.value()
    }
    /// Restore fresh-state behavior while preserving domain configuration.
    pub fn reset(&mut self) {
        self.input.reset();
        self.breakevens = 0;
    }
    /// Return valid observation count.
    pub fn len(&self) -> usize {
        self.input.len()
    }
    /// Return whether no valid observation was processed.
    pub fn is_empty(&self) -> bool {
        self.input.is_empty()
    }
}

crate::impl_observation_metric_lifecycle!(BreakevenRate);
